Framework/Algorithms/src/CalculateFlatBackground.cpp

```diff
@@ -443,5 +499,40 @@ double CalculateFlatBackground::LinearFit(API::MatrixWorkspace_sptr WS,
   return slope * centre + intercept;
 }
 
+/**
+* Utilizes cyclic boundary conditions when calculating the
+* average in the window.
+* @param WS The workspace to operate on
+* @param wsIndex The workspace index to operate on
+* @param windowWidth Width of the averaging window in bins
+* @return Minimum
+*/
+double
+CalculateFlatBackground::movingAverage(API::MatrixWorkspace_const_sptr WS,
+                                       int wsIndex, size_t windowWidth) const {
+  const auto &ys = WS->y(wsIndex);
+  double currentMin = std::numeric_limits<double>::max();
+
+  // Initialize sum.
+  double sum = 0;
+  for (size_t i = 0; i < windowWidth; ++i) {
+    sum += ys[i];
+  }
+  // When moving the window, we need to subtract the single point "falling off"
+  // while adding a new point. Saves us summing all the points in the window.
+  for (size_t i = 0; i < ys.size() - 1; ++i) {
+    currentMin = std::min(currentMin, sum / static_cast<double>(windowWidth));
+    sum -= ys[i];
+    size_t j = i + windowWidth;
+    // Cyclic boundary conditions.
+    if (j >= ys.size()) {
+      j -= ys.size();
+    }
+    sum += ys[j];
```

Comment:
A more stable way to update `sum` would be `sum += ys[j] - ys[i]`. This could prevent warning messages in some edge cases.